tools/test-proxy/Azure.Sdk.Tools.TestProxy/Common/RecordSession.cs

```diff
@@ -103,6 +103,14 @@ public RecordEntry Lookup(RecordEntry requestEntry, RecordMatcher matcher, IEnum
             }
         }
 
+        public void Remove(RecordEntry entry)
+        {
+            lock (Entries)
```

Comment:
If it would improve perf, you might be able to avoid locking by changing `Entries` from `List<RecordEntry>` to `ConcurrentQueue<RecordEntry>`, which would allow you to enqueue/peek/dequeue safely without locking.

Reply:
> If it would improve perf, you might be able to avoid locking by changing `Entries` from `List<RecordEntry>` to `ConcurrentQueue<RecordEntry>`, which would allow you to enqueue/peek/dequeue safely without locking.

I did look at doing this earlier in the lifecycle of this thing. I ran into really wonky matching errors that had to do with needing to effectively "recycle" the queue. Match operations from Task A would _fail_ because an item was popped off the concurrent queue from another lookup operation running from Task B.

The object is synced between the two threads, but one looked in the split second before an item was added back to the queue.

That's why I left it at `lock` originally. I entirely agree that perf wise it won't be as fast (not even close) as a ConcurrentQueue.

Reply:
There's probably a way to make ConcurrentQueue work reliably, but this might not even be a perf bottleneck at all.